File: Algorithms/0113_Path_Sum_II/Python/Path_Sum_II_Solution_1.py

# Space: O(n)
# Time: O(n)

# DFS approach

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def pathSum(self, root, sum):

        def dfs(root, current_path, target_sum, res):
            if root is None: return

            current_path.append(root.val)

            if root.left is None and root.right is None and root.val == target_sum:
                res.append(current_path)

            # caution: make sure to use deep copy(current_path[:]) instead of shallow copy
            #          otherwise the current_path will include all traversed node.
            dfs(root.left, current_path[:], target_sum - root.val, res)
            dfs(root.right, current_path[:], target_sum - root.val, res)

            return res

        return dfs(root, [], sum, [])

File: Algorithms/0113_Path_Sum_II/Java/PathSumIISolution1.java

// Space: O(n)
// Time: O(n)

class Solution {
    public List<List<Integer>> pathSum(TreeNode root, int targetSum) {
        if (root == null) return new ArrayList<>();
        List<List<Integer>> res = new ArrayList<>();
        Queue<List<Object>> queue = new ArrayDeque<>();

        List<Object> temp = new ArrayList<>();
        temp.add(root);
        temp.add(new ArrayList<>(Arrays.asList(root.val)));
        queue.add(temp);

        while (!queue.isEmpty()) {
            List<Object> cur = queue.poll();
            TreeNode curRoot = (TreeNode) cur.get(0);
            ArrayList<Integer> curList = (ArrayList) cur.get(1);
            if (curRoot.left == null && curRoot.right == null) {
                if (curList.stream().mapToInt(Integer::intValue).sum() == targetSum) {
                    res.add(curList);
                }
            }
            if (curRoot.left != null) {
                List<Object> temp1 = new ArrayList<>();
                temp1.add(curRoot.left);
                List<Integer> tempNum = new ArrayList<>(curList);
                tempNum.add(curRoot.left.val);
                temp1.add(tempNum);
                queue.add(temp1);
            }
            if (curRoot.right != null) {
                List<Object> temp1 = new ArrayList<>();
                temp1.add(curRoot.right);
                List<Integer> tempNum = new ArrayList<>(curList);
                tempNum.add(curRoot.right.val);
                temp1.add(tempNum);
                queue.add(temp1);
            }
        }
        return res;
    }
}





